### PR TITLE
Support hadoop-lz4

### PR DESCRIPTION
**Pull request by DorianZheng**
Signed-off-by: dorianzheng <xingzhengde72@gmail.com>

## What type of PR is this：
- [x] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #
https://github.com/StarRocks/starrocks/issues/3008

## Problem Summary(Required) ：

hadoop-lz4 is not compatible with lz4 CLI.
hadoop-lz4 uses a block compression scheme on top of lz4.  As per the hadoop docs
(BlockCompressorStream.java and BlockDecompressorStream.java) the input is split
into blocks.  Each block "contains the uncompressed length for the block followed
by one of more length-prefixed blocks of compressed data."
This is essentially blocks of blocks.
The outer block consists of:
  - 4 byte big endian uncompressed_size
  < inner blocks >
  ... repeated until input_len is consumed ...
The inner blocks have:
  - 4-byte big endian compressed_size
  < lz4 compressed block >
  - 4-byte big endian compressed_size
  < lz4 compressed block >
  ... repeated until uncompressed_size from outer block is consumed ...

the hadoop lz4codec source code can be found here:
https://github.com/apache/hadoop/blob/trunk/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-nativetask/src/main/native/src/codec/lz4codec.cc

More details refer to https://issues.apache.org/jira/browse/HADOOP-12990

### CODE REVIEW

File: gensrc/proto/types.proto [hunk 1]
```diff
@@ -81,5 +81,6 @@ enum CompressionTypePB {
     BZIP2 = 10;
     LZO = 11; // Deprecated
     BROTLI = 12;
+    LZ4_BLOCK = 13;
```
Comment (imay) [inline gensrc/proto/types.proto:84]: Should explain the differences between this and LZ4

File: be/src/util/block_compression.cpp [hunk 2]
```diff
@@ -86,6 +87,78 @@ class Lz4BlockCompression : public BlockCompressionCodec {
...
+public:
+    static const Lz4BlockCompression* instance() {
+        static Lz4BlockCompression s_instance;
+        return &s_instance;
+    }
+
+    ~Lz4BlockCompression() override = default;
+
+    Status compress(const Slice& input, Slice* output) const override {
+        return Status::NotSupported("Lz4BlockCompression::compress not supported yet");
```
Comment (imay) [inline be/src/util/block_compression.cpp:100]: Support compress function then the UT framework can work for this compress type

File: be/src/util/block_compression.cpp [hunk 3]
```diff
@@ -86,6 +87,78 @@ class Lz4BlockCompression : public BlockCompressionCodec {
     size_t max_input_size() const override { return LZ4_MAX_INPUT_SIZE; }
 };
 
+class Lz4BlockCompression : public BlockCompressionCodec {
```
Comment (chaoyli) [inline be/src/util/block_compression.cpp:90]: What's the difference between LZ4 and LZ4Block?
LZ4 is a block algorithm default?

File: be/src/exec/parquet/utils.cpp [hunk 4]
```diff
@@ -10,8 +10,9 @@ CompressionTypePB convert_compression_codec(tparquet::CompressionCodec::type cod
         return NO_COMPRESSION;
     case tparquet::CompressionCodec::SNAPPY:
         return SNAPPY;
+    // parquet-mr uses hadoop-lz4. more details refers to https://issues.apache.org/jira/browse/PARQUET-1878
     case tparquet::CompressionCodec::LZ4:
-        return LZ4;
+        return LZ4_HADOOP;
```
Comment (imay) [inline be/src/exec/parquet/utils.cpp:15]: If this parquet file is not generated by Hadoop, can this file be parsed properly after this PR?

Comment (DorianZheng) [reply be/src/exec/parquet/utils.cpp:15]: Yep, I will fall back to lz4. 

https://github.com/StarRocks/starrocks/pull/3222/files#diff-6d4580b48dea040705eab7a993d69c6ebc36ae8caaeeab3a5e9967abc72da1bfR168

File: be/src/util/block_compression.cpp [hunk 5]
```diff
@@ -86,6 +87,162 @@ class Lz4BlockCompression : public BlockCompressionCodec {
     size_t max_input_size() const override { return LZ4_MAX_INPUT_SIZE; }
 };
 
+// hadoop-lz4 is not compatible with lz4 CLI.
```
Comment (chaoyli) [inline be/src/util/block_compression.cpp:90]: CLI is redundant. 
```
hadoop-lz4 is not compatible with LZ4.
```

Comment (DorianZheng) [reply be/src/util/block_compression.cpp:90]: lz4 CLI means lz4 command line

**Comment by DorianZheng** [comment]
run starrocks_be_unittest

File: be/src/util/block_compression.cpp (hunk 5 above)
Comment (chaoyli) [reply be/src/util/block_compression.cpp:90]: OK

**Review by chaoyli — DISMISSED** [review]

**Review by imay — DISMISSED** [review]
LGTM

**Review by caneGuy — DISMISSED** [review]
other LGTM

**Comment by DorianZheng** [comment]
run starrocks_be_unittest

File: be/src/util/block_compression.cpp [hunk 6]
```diff
@@ -86,6 +87,162 @@ class Lz4BlockCompression : public BlockCompressionCodec {
...
+//   ... repeated until input_len is consumed ...
+// The inner blocks have:
+//   - 4-byte big endian compressed_size
+//   < lz4 compressed block >
+//   - 4-byte big endian compressed_size
+//   < lz4 compressed block >
+//   ... repeated until uncompressed_size from outer block is consumed ...
+//
+// the hadoop lz4codec source code can be found here:
+// https://github.com/apache/hadoop/blob/trunk/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-nativetask/src/main/native/src/codec/lz4codec.cc
```
[the remark this answers is not in the document]
Comment (DorianZheng) [reply be/src/util/block_compression.cpp:108]: https://github.com/apache/hadoop/blob/trunk/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-nativetask/src/main/native/src/codec/Lz4Codec.cc

**Review by caneGuy — APPROVED** [review]

**Comment by DorianZheng** [comment]
run starrocks_be_unittest

**Comment by DorianZheng** [comment]
run starrocks_be_unittest

**Review by chaoyli — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Pull request merged**